feat(revenue-analytics): Revenue Analytics person virtual revenue properties - the empire strikes back

rafaeelaudibert opened the pull request:
This introduces two new properties on the `persons` table: `$virt_revenue` and `$virt_revenue_last_30_days` - both pretty self-explanatory. This is achieved by expanding our tools in several different ways:

- We need to find a way to connect revenue analytics views to the `persons` table. This is achieved by creating a join between the view and the `persons` table (on this direction, DWH team will work on bi-directional joins eventually). This join can be created both on the normal SQL visualization page (no help) or in the Revenue Analytics settings page (we help the customer create that by restricting how that's created).
- There's a new `raw_revenue_analytics_person` table in the backend that handles connecting/grouping the data between a person (from all their distinct ids) to their revenue coming from any revenue source (events or dwh tables).
- There's also a new `.revenue_analytics` accessor on the `persons` table which makes you reach the revenue, e.g. `persons.revenue_analytics.revenue`.
- And, lastly, a simple virtual accessor on the `persons` table that let's you access the properly directly via `persons.$virt_revenue`, which can then be used for filters/grouping in the UI.

We'll implement this for groups in a follow-up PR because this PR is already way too complicated.

This is gated behind the Revenue Analytics FF since it requires the views to exist before the connections can be created.

Closes #35015

rafaeelaudibert committed:
refactor: Remove `modifiers` param

We arent going to use this anymore, let's remove it from everywhere

rafaeelaudibert committed:
refactor(revenue-analytics): Simplify Revenue Analytics view creation

Let's create them straight from the `Team` rather than requiring the `database.py` file to know about the intricacies of how these views exist

rafaeelaudibert committed:
refactor(database): Extract virtual fields initialization into a separate function

Moved the initialization of virtual fields for initial domain and channel types into the `_use_virtual_fields` function to improve code organization and readability. This change reduces duplication and enhances maintainability.

rafaeelaudibert committed:
refactor: Expose `HogQLEditor` `placeholder` prop via `HogQLDropdown`

rafaeelaudibert committed:
refactor: Prevent join tables/json columns from being selected on join

rafaeelaudibert committed:
refactor: Expose custom placeholder on joins modal

rafaeelaudibert committed:
feat: Simplify event view detection

rafaeelaudibert committed:
feat: Introduce virtual person revenue properties

We want people to have access to revenue data on the persons table by doing something like `persons.$virt_revenue`. This achieves this by creating a new lazy table called `raw_persons_revenue_analytics` (we'll have a `groups` one soon) that we can then join on. This implies people can get access to it either via `persons.revenue_analytics.revenue` OR `persons.$virt_revenue`. These can be used for filtering too.

People need to choose how we can find Stripe customers in PostHog by creating a join table between the revenue analytics view AND the `persons` table. We'll have some follow-up commits making that creation slightly easier

rafaeelaudibert committed:
feat: Allow join for managed views

These work just fine, that's not a big deal, let's just create them

rafaeelaudibert committed:
feat(): Expose joins to the frontend

Rather than only exposing our database to the UI, we're now also exposing all of the joins

rafaeelaudibert committed:
feat: Implement "persons join" selector on Revenue Config

We can now more easily create a join between the revenue views and the `persons` table. Rather than letting the user choose all of the tables/fields in existence, we restrict them to reasonable ones.

rafaeelaudibert committed:
fix: Fix DatabaseSchemaQuery mocks

The interface changed, so let's update the mocks to avoid snapshot errors

rafaeelaudibert commented on frontend/__snapshots__/scenes-app-settings-environment--settings-environment-revenue-analytics--dark.png:
Gotta take the time and change the implementation to actually display some external data sources. One of them where the join is set, another one where it isnt.

rafaeelaudibert committed:
fix: Readd events namespace

This was incorrecly removed

Gilbert09 commented on posthog/hogql/ast.py:
```diff
@@ -768,7 +768,7 @@ class ExprCall(Expr):
 class JoinConstraint(Expr):
     expr: Expr
-    constraint_type: Literal["ON", "USING"]
+    constraint_type: Literal["ON", "USING"] = "ON"
```
Hmm, why do we want to set a default here?

rafaeelaudibert replied:
Oh yeah, this is unintended

Gilbert09 reviewed (approved):
I think this is fine, hard to review with so many changes - would appreciate a more detailed PR description next time though please 🙏

rafaeelaudibert committed:
fix: Avoid using DeepPartial

Typegen is completely broken if we use `DeepPartial` in here, let's just use `Partial`

rafaeelaudibert committed:
fix: Fix mypy types

rafaeelaudibert committed:
fix: Fix base view tests

rafaeelaudibert committed:
fix: Add missing `events`

This changed names, so let's update it

rafaeelaudibert committed:
feat: Add joins and source to rev analytics settings snapshot

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Add missing return

rafaeelaudibert committed:
fix: Add missing `events` namespace

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

rafaeelaudibert commented on products/revenue_analytics/frontend/settings/ExternalDataSourceConfiguration.tsx:
```diff
@@ -59,6 +64,57 @@ export function ExternalDataSourceConfiguration({
+                            const itemPrefix = item.prefix
+                                ? `${item.source_type.toLowerCase()}.${item.prefix.replace(/_+$/, '')}`
+                                : item.source_type.toLowerCase()
+                            const joinName = `${itemPrefix}.customer_revenue_view`
```
Yeah, will need to tweak this eventually, but this works for now.

rafaeelaudibert committed:
fix: Small fixes

robbie-c commented on posthog/hogql/database/database.py:
```diff
@@ -371,6 +371,40 @@ def _use_error_tracking_issue_id_from_error_tracking_issue_overrides(database: D
+            properties_path=["poe", "properties"],
+        )
+
+    # TODO: POE is not well supported yet, that part is a stub
```
Probably yes, but do what you think will confuse people the least. Customers can use pdi.properties.$virt_revenue if they want to grab the latest version of this property

robbie-c replied:
Copy this test, which runs in all 4 modes https://github.com/PostHog/posthog/blob/472b6a5042fdbf6c7daed7d85b113caad4652b09/posthog/hogql/database/schema/test/test_persons.py#L248

robbie-c reviewed (approved):
LGTM, with caveats
* @Gilbert09's comments look legit
* This is pretty hard to review due to the size
* This is all behind a FF

rafaeelaudibert committed:
refactor: Remove unintended change

I wanna change this eventually, but dont wanna do it now

rafaeelaudibert committed:
refactor: Fix mypy checks

rafaeelaudibert committed:
tests: Prepare tests for future Rev Analytics POE code

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert commented:
Back in draft, meeting with Robbie tomorrow to solve poe

rafaeelaudibert committed:
fix: Fix missing `.events` key

rafaeelaudibert committed:
fix: Fix incorrect import

rafaeelaudibert committed:
fix: fix mypy

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (3)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update query snapshots

rafaeelaudibert committed:
fix: Fix missing timestamp freeze

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)
